crates/compiler/unify/src/unify.rs

```diff
@@ -340,9 +342,21 @@ impl<'a> Env<'a> {
     }
 }
 
+/// Unifies two types.
+/// The [mode][Mode] enables or disables certain extensional features of unification.
+///
+/// `observed_pol` describes the [polarity][Polarity] of the type observed to be under unification.
+/// This is only relevant for producing error types, and is not material to the unification
+/// algorithm.
 #[inline(always)]
-pub fn unify(env: &mut Env, var1: Variable, var2: Variable, mode: Mode) -> Unified {
-    unify_help(env, var1, var2, mode)
+pub fn unify(
+    env: &mut Env,
+    var1: Variable,
+    var2: Variable,
+    mode: Mode,
+    observed_pol: Polarity,
+) -> Unified {
+    unify_help(env, var1, var2, mode, observed_pol)
 }
```

Comment:
Kind of unfortunate to force unification to care about polarity solely for the purpose of generating error types. In a refactor we can pull out generation of error types, or abstract it.